### PR TITLE
frontend/#4118/D2D and D2C referral flow with deeplink

### PR DESCRIPTION
**Pull request by ashishsingh101**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

1. driver can refer another driver by sharing referal playstore link for cases when app is not installed
2. referral through deeplink also supported for cases when app is already installed

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Frontend/android-native/app/src/user/common/res/drawable/ny_ic_referral_already_applied.png
Comment (Rohit4997) [inline Frontend/android-native/app/src/user/common/res/drawable/ny_ic_referral_already_applied.png:1]: @ashishsingh101  try to reduce image size

File: Frontend/android-native/app/src/user/common/res/drawable/ny_ic_referral_invalid.png
Comment (Rohit4997) [inline Frontend/android-native/app/src/user/common/res/drawable/ny_ic_referral_invalid.png:1]: @ashishsingh101  try to reduce image size < 10KB

File: Frontend/android-native/app/src/driver/nammaYatriPartner/AndroidManifest.xml [hunk 1]
```diff
@@ -48,6 +48,7 @@
                 <data android:scheme="http"/>
                 <data android:host="nammayatri.in"/>
                 <data android:path="/p" />
+                <data android:path="/refer" />
```
Comment (Rohit4997) [inline Frontend/android-native/app/src/driver/nammaYatriPartner/AndroidManifest.xml:51]: This can be added in main AndroidManifeast

File: Frontend/android-native/app/src/user/nammaYatri/AndroidManifest.xml [hunk 2]
```diff
@@ -22,6 +22,7 @@
                 <data android:scheme="http"/>
                 <data android:host="nammayatri.in"/>
                 <data android:path="/u" />
+                <data android:path="/refer" />
```
Comment (Rohit4997) [inline Frontend/android-native/app/src/user/nammaYatri/AndroidManifest.xml:25]: This can be added in main AndroidManifeast

File: Frontend/ui-common/src/Engineering/Helpers/Utils.purs [hunk 3]
```diff
@@ -333,3 +333,24 @@ fetchLanguage currLang = case currLang of
...
                   
+getReferralCode :: String -> Maybe String
+getReferralCode referralData =
+  let tuples = parseKeyValues referralData "&" "="
+  in findValueFromTuples tuples "utm_campaign"
+
+parseKeyValues :: String -> String -> String -> Array (Tuple String String)
+parseKeyValues string queryPattern keyValuePattern = DA.catMaybes $ map (\query -> parseKeyValueWithPattern query keyValuePattern) (split (Pattern queryPattern) string)
+
+parseKeyValueWithPattern :: String -> String -> Maybe (Tuple String String)
```
Comment (Rohit4997) [inline Frontend/ui-common/src/Engineering/Helpers/Utils.purs:344]: `parseKeyValueWithPattern :: String -> String -> Maybe (Tuple String String)
parseKeyValueWithPattern string pattern = 
  case split (Pattern pattern) string of
    [key, value] -> Just (Tuple key value)
    _ -> Nothing`
 
 Try this simplified version

File: Frontend/ui-driver/src/Flow.purs [hunk 4]
```diff
@@ -335,8 +335,11 @@ enterOTPFlow = do
       setValueToLocalStore DRIVER_ID driverId
       void $ liftFlowBT $ setCleverTapUserData "Identity" (getValueToLocalStore DRIVER_ID)
       setValueToLocalStore REGISTERATION_TOKEN resp.token -- add from response
-      void $ lift $ lift $ toggleLoader false
+      if (getValueToLocalStore REFERRER_URL) == "__failed" then
```
Comment (Rohit4997) [inline Frontend/ui-driver/src/Flow.purs:338]: use when here

File: Frontend/ui-customer/src/Flow.purs [hunk 5]
```diff
@@ -241,10 +241,34 @@ currentFlowStatus = do
...
         void $ lift $ lift $ Remote.updateProfile (UpdateProfileReq requiredData)
     
-    updateUserLanguage :: Maybe String -> FlowBT String Unit
-    updateUserLanguage language = 
-      when (isNothing language || (getKeyByLanguage (fromMaybe "ENGLISH" language) /= (getLanguageLocale languageKey)))
-        $ void $ lift $ lift $ Remote.updateProfile (Remote.mkUpdateProfileRequest FunctionCall)
+    updateLanguageAndReferralCode :: Maybe String -> FlowBT String Unit
+    updateLanguageAndReferralCode language = do
+      config <- getAppConfigFlowBT appConfig
+      let referralStatus = getValueToLocalStore REFERRAL_STATUS
```
Comment (Rohit4997) [inline Frontend/ui-customer/src/Flow.purs:247]: Can you mention the usage of this local storage variable and
What all values it can have ?

File: Frontend/ui-customer/src/Main.purs [hunk 6]
```diff
@@ -99,13 +100,41 @@ onConnectivityEvent triggertype = do
...
+  case payload of
+    Right payload'  -> do
+        mainFiber <- launchAff $ flowRunner defaultGlobalState $ do
+          _  <- case (runFn2 JBridge.getMainFiber Just Nothing) of
+            Nothing -> pure unit
+            Just fiber -> liftFlow $ launchAff_ $ killFiber (error "error in killing fiber") fiber
+          _ <- runExceptT $ runBackT do
+                  case event.type of 
+                    "REFERRAL" -> do
+                      setValueToLocalStore REFERRER_URL event.data
```
Comment (Rohit4997) [inline Frontend/ui-customer/src/Main.purs:125]: Instead of using local storage variable `REFERRER_URL`, Can we send event.data in payload of baseAppFlow ?

File: Frontend/ui-driver/index.js [hunk 7]
```diff
@@ -211,13 +219,19 @@ window.onMerchantEvent = function (_event, payload) {
       }else if (parsedPayload.payload.notificationData && parsedPayload.payload.notificationData.notification_type == "PAYMENT_MODE_MANUAL") {
         purescript.main(makeEvent("PAYMENT_MODE_MANUAL", ""))();
       } else if (parsedPayload.payload.viewParam){
-        purescript.onNewIntent(makeEvent("DEEP_VIEW", parsedPayload.payload.viewParam))();
+        if (!checkForReferral(parsedPayload.payload.viewParam, "REFERRAL_NEW_INTENT")) {
```
Comment (Rohit4997) [inline Frontend/ui-driver/index.js:222]: Can we write it in a different function, this `onMerchantEvent` function looks very messy ?

File: Frontend/ui-driver/src/Flow.purs [hunk 8]
```diff
@@ -476,14 +479,23 @@ onBoardingFlow = do
   logField_ <- lift $ lift $ getLogFields
   _ <- pure $ hideKeyboardOnNavigation true
   config <- getAppConfigFlowBT Constants.appConfig
-  globalstate <- getState 
+  let cityConfig = getCityConfig config.cityConfig (getValueToLocalStore DRIVER_LOCATION)
+  (GlobalState allState) <- getState 
+  let registrationState = allState.registrationScreen
   permissions <- checkAllPermissions false config.permissions.locationPermission
-  (GetDriverInfoResp getDriverInfoResp) <- getDriverInfoDataFromCache globalstate false
+  (GetDriverInfoResp getDriverInfoResp) <- getDriverInfoDataFromCache (GlobalState allState) false
```
Comment (Rohit4997) [inline Frontend/ui-driver/src/Flow.purs:486]: Why this change? It's not required I think

File: Frontend/ui-driver/src/Flow.purs [hunk 9]
```diff
@@ -3158,4 +3161,31 @@ driverEarningsFlow = do
...
+          logField_ <- lift $ lift $ getLogFields
+          liftFlowBT $ logEventWithMultipleParams logField_ "ny_driver_ref_code_submitted" $ [{key : "Referee Code", value : unsafeToForeign code}]
+          void $ pure $ setCleverTapUserProp [{key : "Entered Referral Code", value : unsafeToForeign code}] 
+          pure $ toast "Referral Code Appiled"
+        _         -> do
+          void $ pure $ cleverTapCustomEvent "ny_driver_ref_code_already_applied"
+          pure $ toast "Referral Code Already Appiled"
+      referralCodeAppliedStage state
+    Left errorPayload -> do
+      pure $ toast "Incorrect Referral Code"
```
Comment (Rohit4997) [inline Frontend/ui-driver/src/Flow.purs:3182]: Add translations

File: Frontend/ui-driver/src/Flow.purs [hunk 10]
```diff
@@ -3158,4 +3161,31 @@ driverEarningsFlow = do
...
+activateReferralCode state code = do
+  referDriverResponse <- lift $ lift $ Remote.referDriver $ makeReferDriverReq code
+  case referDriverResponse of
+    Right (ReferDriverResp (API.ApiSuccessResult res)) -> do
+      case res.result of
+        "Success" -> do
+          logField_ <- lift $ lift $ getLogFields
+          liftFlowBT $ logEventWithMultipleParams logField_ "ny_driver_ref_code_submitted" $ [{key : "Referee Code", value : unsafeToForeign code}]
+          void $ pure $ setCleverTapUserProp [{key : "Entered Referral Code", value : unsafeToForeign code}] 
+          pure $ toast "Referral Code Appiled"
```
Comment (Rohit4997) [inline Frontend/ui-driver/src/Flow.purs:3176]: Add translations

File: Frontend/ui-driver/src/Flow.purs [hunk 11]
```diff
@@ -3158,4 +3161,31 @@ driverEarningsFlow = do
...
+        _         -> do
+          void $ pure $ cleverTapCustomEvent "ny_driver_ref_code_already_applied"
+          pure $ toast "Referral Code Already Appiled"
```
Comment (Rohit4997) [inline Frontend/ui-driver/src/Flow.purs:3179]: Add translations here also

**Review by Rohit4997 — CHANGES_REQUESTED** [review]

File: Frontend/android-native/app/src/driver/nammaYatriPartner/AndroidManifest.xml (hunk 1 above)
Comment (ashishsingh101) [reply Frontend/android-native/app/src/driver/nammaYatriPartner/AndroidManifest.xml:51]: no it can't be because of different domains

File: Frontend/android-native/app/src/user/nammaYatri/AndroidManifest.xml (hunk 2 above)
Comment (ashishsingh101) [reply Frontend/android-native/app/src/user/nammaYatri/AndroidManifest.xml:25]: it can't be

File: Frontend/android-native/app/src/driver/nammaYatriPartner/AndroidManifest.xml (hunk 1 above)
Comment (Rohit4997) [reply Frontend/android-native/app/src/driver/nammaYatriPartner/AndroidManifest.xml:51]: What do you mean by different domains ?

**Review by Rohit4997 — APPROVED** [review]

**Review by witcher-shailesh — APPROVED** [review]

**Review by witcher-shailesh — APPROVED** [review]
